### PR TITLE
Add option to force in-order answering of questions

### PR DESCRIPTION
**Pull request by mwest1066**
This started life as #2626, but ultimately that had to be reverted due to deployment issues. Now that https://github.com/PrairieLearn/PrairieLearn/pull/4412 has been merged and deployed, we can try it again.

The original list of commits, for posterity:

* Sync strict sequence rules to db from assessmentInfo

* Fixing HTML in studentAssessmentInstanceHomework.ejs

* Displaying strict sequence rules in instructor and student views

* Disabling links on student homework questions page in strict sequence

* Cleaning up student assessments html

* Renaming migration appropriately

* Changing to nested sequence structure in assessment json

* Changing threshold name, value format

* Fixing sync typo

* Adding sproc to check for sequence block, enforce naming conventions

* Updating student view pages to respond to sproc return values

* Add distinction between end of assessment and sequence blocked questions

* Fix linter errors

* Updating zones.pg to reflect database changes

* Blocking access server-side to sequence blocked instance questions

* Implementing forceSequential UI changes for student exam views

* Rename 404_zones__sequence_params.sql to 179_zones__sequence_params.sql

:P

* Update ChangeLog.md

* adding assessment

* Bump migration to 180

* change to assessment

* Bump migration to 181

* draft newton questions

* changing assessment file

* last updates to question

* Fix zone sequence values not updating on sync

* Clamping scorePercThreshold on sync

* Now syncing implicit, multi-level sequence rules:
assessment, zone, alternative group, and question levels

* Updating WS01 to follow new format

* Splitting sproc in two, some of the functionality is needed elsewhere

* Integrating changes into UI: Student assessment pages

* Removing old sequence info from instructorAssessmentQuestions

* Removing old sequence info from authz

* Removing old sequence info from question page

* Checking for best submission score percentage instead of overall score

* Adding an extra sequence parameter to WS1 for better interactive example

* Refactor backend to use symbol 'min_advance_perc', adding null check

* Refactor sync to use symbol 'minAdvancePerc'

* Refactoring find_unlock_score_perc to take an assessment question id

* Adding column to instructorAssessmentQuestions for minAdvancePerc

* Update logic/wording in sproc

* Dropping "block" in favor of "lock" to avoid user access confusion

* Alphabetizing columns in .pg files

* Fix out-of-order columns in .pg file

* Adding exception to continue in sequence when question not open

* Fix merge conflict

* Documentation first draft

* Adding popover text for "Previous question" button

* Adding ids/classes to html for testing

* Adding initial JS tests

* JS Linter (it's late don't judge me)

* Indirectly testing for zone sync

* Merge branch 'master' into forceSequential

* Removing demoNewton redundancy

* Fixing typo / unnecessary with() in middleware

* Fixing sync issues overlooked in merge

* Adding minAdvancePerc to typedefs

* Adding warning header to docs page

* Switching test exam to sentence case

* Renaming worksheet folder in exampleCourse

* Removing redundant aset

* Moving minAdvancePerc column in instructorAssessmentQuestions

* Calculating effective_min_advance_perc at sync time

* Removing redundant sprocs

* Finish block -> lock refactor

* Calculate sequence_locked inside of question_order

* Add "if exists" to drop function

* Update database/tables

* Linter fix

* When a question locks the next question, it locks all questions after

* Updating tests to reflect new logic

* Refactor usage of 'minAdvancePerc' to 'advanceScorePerc'

* Refactor pt 2

* Alphabetize db columns

* Fix conflict wuth new test in master

* Change docs wording per Matt's request

* No honor code in worksheetTemplate

* Only show Advance Score column if feature is utilized

* Revert style changes in instructorAssessmentQuestions zone header

* No abbreviated clause

* Updated question_order for new behavior, untested

* Add lock icon, change colors back to originals

* Fix confusing wording on student ai page, move to partial

* Squashed merge with master

* Move forceSequential tests to Exam 11

* No redundant column names

* Remove unnecessary JSON entries for prev iq

* Fix undeclared alias usage

* Revert cosmetic changes unrelated to PR

* Update sequential tests to account for cascade unlocking questions

* WIP News Item

* Pass JS linter

* Complete news item

* Test extra hard that interface matches expected values

* Fix incorrect worksheet number

* Update info.json

we usually have another person releasing the news (and not the code developer)

* Move demo pages to individual zones

* Reverse accidental committed timeout change

* Fix typo

* Renumber and edit news item

* Renumber migration

* Fix news_items and migrations numbering

* rename variable in instructorAssessmentQuestions

* Use full advance_score_perc name

* Specify percentage in code

* Don't rename advance_score_perc

* Add file attachment popover in question label

* Remove redundant popover initializer

* Matt's style requests

* Make history by nullishly coallescing

* Remove unused score_perc value for next question

* Descriptive variable names instead of next/prev

* Bump news item index

* Linter doesn't like the nullish coallesce operator

This reverts commit 20ac30b41b517274a523ac9615bca8244991845f.

* Reintroduce nullish coallesce operator

This reverts commit f79fe9f64095b93353e9e1292d28d017f08a7fb3.

* Fix typo: +be

* Clarify wording for sequence unlocking

* Include alternative advanceScorePerc in assessment sync

* Update eslint to be in-line with master

Co-authored-by: David Mitchell <david@astrosticks.com>
Co-authored-by: Mariana Silva <mfsilva@illinois.edu>
Co-authored-by: Matthew West <mwest@illinois.edu>

### CODE REVIEW

**Comment by mwest1066** [comment]
@astrosticks this PR is a rebased copy of #2626 that I had applied and then reverted, due to the DB upgrade issues. Those should now be fixed by #4412, so we can try merging/deploying this PR again.

There were a few minor conflicts when I rebased this PR and other things have changed in the meantime with the `master` branch. Could you test this again and make sure that it's fully working before I merge it? Thanks!

**Comment by mwest1066** [comment]
@astrosticks there is a test failure in the in-order tests, which I did not attempt to debug. I'll leave that for you to fix, if that's ok?

**Comment by mwest1066** [comment]
@astrosticks what's the status of your testing for this PR? Did you get a chance to thoroughly test it manually yet?

**Comment by astrosticks** [comment]
@mwest1066 Yes, I tested manually and everything seemed to be working. Still in process of getting "next question" button to update when external grading finishes.

**Comment by mwest1066** [comment]
> @mwest1066 Yes, I tested manually and everything seemed to be working. Still in process of getting "next question" button to update when external grading finishes.

Oh right! I forgot about that remaining issue. Thanks for the reminder!

**Comment by SethPoulsen** [comment]
Is there an arrival estimate for this? We were hoping to use this for a study we are doing about mid-semester. Thanks!

**Comment by mwest1066** [comment]
> Is there an arrival estimate for this? We were hoping to use this for a study we are doing about mid-semester. Thanks!

No specific target. It's on my list of PRs to review. You know how to poke me to get it done 😄

**Comment by jasonxia17** [comment]
> No specific target. It's on my list of PRs to review. You know how to poke me to get it done

We'd also love to have this feature in 374 as soon as possible, so it would be great if you could prioritize reviewing this PR! :slightly_smiling_face:  We've already released two guided problem sets where later questions give away some answers to earlier questions, and we anticipate that most of our weekly problem sets will be designed in this manner.

File: sprocs/sync_assessments.sql [hunk 1]
```diff
@@ -269,39 +270,46 @@ BEGIN
...
                 INSERT INTO alternative_groups (
                     number,
                     number_choose,
+                    advance_score_perc,
                     assessment_id,
                     zone_id
                 ) VALUES (
                     (alternative_group->>'number')::integer,
                     (alternative_group->>'number_choose')::integer,
+                    (alternative_group->>'advance_score_perc')::integer,
```
Comment (nwalters512) [inline sprocs/sync_assessments.sql:305]: I believe this should be casted as a `double precision`?

```suggestion
                    (alternative_group->>'advance_score_perc')::double precision,
```

File: exampleCourse/courseInstances/Sp15/assessments/ws01-worksheetTemplate/infoAssessment.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        {
+            "questions": [
+                {"id": "demo/demoNewton-page1", "points": [5,2], "advanceScorePerc": 50}
+            ]
+        },
+        {
+            "questions": [
+                {"id": "demo/demoNewton-page2", "points": [5,4,3,2,1]}
+            ]
+        }
```
[the remark this answers is not in the document]
Comment (echuber2) [reply exampleCourse/courseInstances/Sp15/assessments/ws01-worksheetTemplate/infoAssessment.json:24]: I assume the zoning was done to ensure page 1 comes before page 2, although the new docs discourage using the proposed feature for Exam items.

> Within each zone the question order is randomized for Exam assessments.
> \- PL docs

[the remark this answers is not in the document]
Comment (nwalters512) [reply exampleCourse/courseInstances/Sp15/assessments/ws01-worksheetTemplate/infoAssessment.json:24]: Ah, great catch - yeah, we should change this example to be a homework, not an exam. It's perpetually confusing to me to have the two assessment types 😓

File: pages/studentAssessmentInstanceExam/studentAssessmentInstanceExam.ejs [hunk 3]
```diff
@@ -205,16 +207,11 @@
...
-            <tr>
+            <tr class="<% if (instance_question.sequence_locked) { %>bg-light pl-sequence-locked<% } %>">
               <td>
-                <a href="<%= urlPrefix %>/instance_question/<%= instance_question.id %>/">
-                  Question <%= instance_question.question_number %>
-                </a>
-                <% if (instance_question.file_count > 0) { %>
-                  <i class="fas fa-paperclip"></i>
-                <% } %>
+                <% instance_question.rowLabelText = `Question ${instance_question.question_number}`;%>
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply pages/studentAssessmentInstanceExam/studentAssessmentInstanceExam.ejs]: Resolved in 22be867.

File: pages/studentAssessmentInstanceExam/studentAssessmentInstanceExam.ejs [hunk 4]
```diff
@@ -161,7 +161,7 @@
           <% } %>
         </div>
 
-        <table class="table table-sm table-hover">
+        <table id="assessment-questions" class="table table-sm table-hover">
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply pages/studentAssessmentInstanceExam/studentAssessmentInstanceExam.ejs]: Resolved in 6a2fe00

File: pages/instructorAssessmentQuestions/instructorAssessmentQuestions.ejs [hunk 5]
```diff
@@ -112,6 +122,11 @@
                   <% if (assessment.type == "Exam") { %><%= question.points_list %><% } %>
                   <% if (assessment.type == "Homework") { %><%= question.init_points %>/<%= question.max_points %><% } %>
                 </td>
+                <% if (showAdvanceScorePercCol) { %>
+                  <td class="pl-sequence-prev-unlock-score <% if (question.assessment_question_advance_score_perc == 0) { %>text-muted<% } %>">
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply pages/instructorAssessmentQuestions/instructorAssessmentQuestions.ejs]: Resolved in 6a2fe00.

File: lib/question.js [hunk 6]
```diff
@@ -2209,6 +2210,37 @@ module.exports = {
               callback(null);
             });
           },
+          (callback) => {
+            // only render if variant is part of assessment
+            if (variant.instance_question_id == null) {
+              return callback(null);
+            }
+
+            // Render the next question nav link
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply lib/question.js:2219]: Resolved in 401c689.

**Comment by mwest1066** [comment]
@nwalters512 I think this PR is ready to merge?

File: news_items/036_sequential_questions/info.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{
+    "uuid": "5b3aaaa8-040f-4ea8-81d8-2dcc7dbf8cff",
+    "title": "Forcing students to answer questions in order",
+    "author": "<a href=\"https://github.com/mfsilva22\">Mariana Silva</a>",
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply news_items/036_sequential_questions/info.json:4]: Ah, that makes sense! I'm happy to stick with that - thanks for explaining.

**Comment by nwalters512** [comment]
> @nwalters512 I think this PR is ready to merge?

@mwest1066 almost - the last thing I want to change is changing the example assessment (`exampleCourse/courseInstances/Sp15/assessments/ws01-worksheetTemplate/infoAssessment.json`) to a homework so that we exemplify our own best practice of "don't force users to answer in order on an exam".

**Comment by mwest1066** [comment]
> > @nwalters512 I think this PR is ready to merge?
> 
> @mwest1066 almost - the last thing I want to change is changing the example assessment (`exampleCourse/courseInstances/Sp15/assessments/ws01-worksheetTemplate/infoAssessment.json`) to a homework so that we exemplify our own best practice of "don't force users to answer in order on an exam".

Oh yes, makes sense -- good catch!

File: exampleCourse/courseInstances/Sp15/assessments/ws01-worksheetTemplate/infoAssessment.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply exampleCourse/courseInstances/Sp15/assessments/ws01-worksheetTemplate/infoAssessment.json]: I changed this to a homework in 778106e and 121911d.

**Review by nwalters512 — APPROVED** [review]
@mwest1066 I'd appreciate a final quick review from you, but I'm feeling good about this!

**Comment by nwalters512** [comment]
The last CI jobs were running without finishing, so I cancelled them. I'll have to dig into that before we merge this 😢

**Comment by mwest1066** [comment]
@nwalters512 lgtm (I can't formally approve because I'm officially the PR author)

**Pull request merged**